File: Scope.py
from typing import Optional, List
from Location import Location
from Symbol import Symbol
from Error import CompilerError
from SymbolTable import SymbolTable
from Error import InternalError, getCallerLocation
from antlr4 import ParserRuleContext


class Statement:
    def __init__(self, ctx: ParserRuleContext):
        self.ctx = ctx


class Scope:
    def __init__(self, location: Location, parentScope: Optional["Scope"]):
        self.symbolTable = SymbolTable()
        self.parentScope = parentScope
        self.terminated = False
        self.location = location
        self.statements: List[Statement] = []

    def defineSymbol(self, symbol: Symbol, loc: Location):
        self.symbolTable.insert(symbol, loc)

    def tryLookupSymbol(self, name: str, loc: Location) -> Optional[Symbol]:
        symbol = self.symbolTable.tryLookup(name, loc)
        if symbol:
            return symbol

        if self.parentScope:
            return self.parentScope.tryLookupSymbol(name, loc)
        return None

    def lookupSymbol(self, name: str, loc: Location):
        symbol = self.symbolTable.tryLookup(name, loc)
        if symbol:
            return symbol
        if self.parentScope:
            return self.parentScope.lookupSymbol(name, loc)
        raise CompilerError(f"Symbol '{name}' was not declared in this scope", loc)

    def mutabilityString(self, symbol: Symbol):
        return "mutable" if symbol.isMutable() else "const"

    def setTerminated(self, terminated: bool):
        self.terminated = terminated

    def isTerminated(self):
        return self.terminated

    def __str__(self):
        return f"Scope({self.location}):\n{self.symbolTable}"

    def __repr__(self):
        return self.__str__()

File: src/Scope.ts
import { CompilerError, type SourceLoc } from "./Errors";
import type { Module } from "./Module";
import type { ParsedSymbol } from "./ParsedTypes";
import type { Statement } from "./Statement";
import type { ScopeId } from "./store";

export class ResolvedScope {
  private symbols: ParsedSymbol.Symbol[] = [];
  public statements: Statement[] = [];
  public terminated: boolean = false;

  constructor(
    public id: ScopeId,
    public location: SourceLoc,
    public parentScope?: ScopeId,
  ) {}

  defineSymbol(symbol: ParsedSymbol.Symbol) {
    if (symbol.variant === "Constant") {
      throw new CompilerError(
        `Cannot define a constant symbol`,
        symbol.location,
      );
    }
    if (this.tryLookupSymbolHere(symbol.name)) {
      throw new CompilerError(
        `Symbol '${symbol.name}' was already defined in this scope`,
        symbol.location,
      );
    }
    this.symbols.push(symbol);
  }

  getAllSymbols() {
    return this.symbols;
  }

  tryLookupSymbol(
    module: Module,
    name: string,
    loc: SourceLoc,
  ): ParsedSymbol.Symbol | undefined {
    let symbol = this.symbols.find((s) => "name" in s && s.name === name);
    if (symbol) {
      return symbol;
    }

    if (this.parentScope) {
      return module.parsedStore
        .getScope(this.parentScope)
        .tryLookupSymbol(module, name, loc);
    }
    return undefined;
  }

  tryLookupSymbolHere(name: string): ParsedSymbol.Symbol | undefined {
    return this.symbols.find((s) => "name" in s && s.name === name);
  }

  lookupSymbol(
    module: Module,
    name: string,
    loc: SourceLoc,
  ): ParsedSymbol.Symbol {
    let symbol = this.tryLookupSymbol(module, name, loc);
    if (symbol) {
      return symbol;
    }
    if (this.parentScope) {
      return module.parsedStore
        .getScope(this.parentScope)
        .lookupSymbol(module, name, loc);
    }
    throw new CompilerError(
      `Symbol '${name}' was not declared in this scope`,
      loc,
    );
  }
}
